Source/Parsing/Frame/Content/Operation/FrameParser.swift

```diff
@@ -46,4 +46,56 @@ extension FrameParser {
         return rolePersonArray as! [(String, String)]
     }
     
+    internal func extractChapterElements(
+        from frameData: inout Data.SubSequence,
+        encoding: ID3StringEncoding,
+        id3Tag: ID3Tag,
+        frameSizeParser: FrameSizeParser,
+        id3FrameParser: ID3FrameParser,
+        frameContentParsingOperationFactory: ID3FrameContentParsingOperationFactory
+    ) -> (
+        elementID: String,
+        startTime: Int,
+        endTime: Int,
+        startByteOffset: Int,
+        endByteOffset: Int,
+        embeddedSubframes: [FrameName: ID3Frame]?
+        ) {
+            let elementID = frameData.extractPrefixAsStringUntilNullTermination(encoding)
+ 
+            let startTimeBytes = frameData.extractFirst(4)
+            let startTimeInt = // need reverse of UInt32ToByteArrayAdapter
+
+            let endTimeBytes = frameData.extractFirst(4)
+
+            let startByteOffsetBytes = frameData.extractFirst(4)
+
+            let endByteOffsetBytes = frameData.extractFirst(4)
+            
+            var embeddedSubframes: [FrameName: ID3Frame]? = [:]
+            let parsedSubframes = parse(
+                embeddedSubframes: frameData,
+                id3Tag: id3Tag,
+                frameSizeParser: frameSizeParser,
+                id3FrameParser: id3FrameParser)
+            // how do I get a [FrameName: ID3Frame] return from this?
+            
+            return (elementID, startTime, endTime, startByteOffset, endByteOffset, embeddedSubframes)
+    }
+  
+    func parse(embeddedSubframes: Data, id3Tag: ID3Tag, frameSizeParser: FrameSizeParser, id3FrameParser: ID3FrameParser) {
+        var currentFramePosition = 0
+        while currentFramePosition < embeddedSubframes.endIndex {
+            let frameSize = frameSizeParser.parse(
+                mp3: embeddedSubframes as NSData,
+                framePosition: currentFramePosition,
+                version: id3Tag.properties.version)
+            let frame = embeddedSubframes.subdata(with: NSMakeRange(currentFramePosition, min(frameSize, embeddedSubframes.endIndex - currentFramePosition)))
```

Comment:
This is basically the `parse` method from `ID3FramesParser`, just tweaked slightly to use the `embeddedSubframes` instead of an mp3 file as the source of the data.

However, I can't quite get this portion with `NSMakeRange` right.

I'm not even sure if I should be parsing the subframes here, or in the `ChapterFrameContentParsingOperation`. I'm getting turned around in my head with all these different parsers, as per usual.

Reply:
That’s an old clumsy Objective C API. You shouldn’t need `NSRange`. You can just get a slice the Swift way with `embeddedSubframes[someStartPosition..<someEndPosition]`

Reply:
Like this?

```swift
            let frame = embeddedSubframes[currentFramePosition..<currentFramePosition+frameSize]
```

Reply:
Yes, except you’ll want to make sure the indices are valid before you try to use them. It will crash if you apply them and they are out of bounds.

Reply:
I thought that was what the portion of the method dealing with the frame size parser did?

Here's the whole thing so far:

```swift
    func parse(embeddedSubframes: Data, id3Tag: ID3Tag, frameSizeParser: FrameSizeParser, id3FrameParser: ID3FrameParser) {
        var currentFramePosition = 0
        while currentFramePosition < embeddedSubframes.endIndex {
            let frameSize = frameSizeParser.parse(
                mp3: embeddedSubframes as NSData,
                framePosition: currentFramePosition,
                version: id3Tag.properties.version)
            let frame = embeddedSubframes[currentFramePosition..<currentFramePosition+frameSize]
            id3FrameParser.parse(frame: frame, frameSize: frameSize, id3Tag: id3Tag)
            currentFramePosition += frame.count;
        }
    }
```

So it shouldn't go beyond the bounds of `embeddedSubframes.endIndex`, correct?

The other thing tripping me up at this point is how I should handle what comes out of the parse method. I've got this method:

```swift
    internal func extractChapterElements(
        from frameData: inout Data.SubSequence,
        encoding: ID3StringEncoding,
        id3Tag: ID3Tag,
        frameSizeParser: FrameSizeParser,
        id3FrameParser: ID3FrameParser,
        frameContentParsingOperationFactory: ID3FrameContentParsingOperationFactory
    ) -> (
        elementID: String,
        startTime: Int,
        endTime: Int,
        startByteOffset: Int,
        endByteOffset: Int,
        embeddedSubframes: [FrameName: ID3Frame]?
        ) {
            let elementID = frameData.extractPrefixAsStringUntilNullTermination(encoding)
 
            let startTime = Int(frameData.extractFirst(4).uint32)
            let endTime = Int(frameData.extractFirst(4).uint32)
            let startByteOffset = Int(frameData.extractFirst(4).uint32)
            let endByteOffset = Int(frameData.extractFirst(4).uint32)
            
            var embeddedSubframes: [FrameName: ID3Frame]? = [:]
            let parsedSubframes = parse(
                embeddedSubframes: frameData,
                id3Tag: id3Tag,
                frameSizeParser: frameSizeParser,
                id3FrameParser: id3FrameParser)
            // how do I get a [FrameName: ID3Frame] return from this?
            
            return (elementID: elementID,
                    startTime: startTime,
                    endTime: endTime,
                    startByteOffset: startByteOffset,
                    endByteOffset: endByteOffset,
                    embeddedSubframes: embeddedSubframes)
    }
```

which is going to be used in this method:

```swift
    func parse(frame: Data, version: ID3Version, completed: (FrameName, ID3Frame) -> ()) {
        var parsing = frame[...]
        extractHeader(from: &parsing, version: version)
        
        let encoding = extractEncoding(from: &parsing)
        
        let parsed = extractChapterElements(
            from: &parsing,
            encoding: encoding)

        let constructed = ID3ChapterFrame(
            elementID: parsed.elementID,
            startTime: parsed.startTime,
            endTime: parsed.endTime,
            startByteOffset: parsed.startByteOffset,
            endByteOffset: parsed.endByteOffset,
            embeddedSubframes: <#T##[FrameName : ID3Frame]#>
        )
        completed(frameName, constructed)
    }
```

I'm just not sure at which point I should be handling the parsing of the embedded subframes.

Reply:
> I thought that was what the portion of the method dealing with the frame size parser did?

Well, I didn’t look at its implementation of `frameSizeParser.parse(...)`, but it sounds like that only parses out how big the frame declares itself to be.

I could take a 300 page book, tear out the last 150 pages and hand it to you. The table of contents at the front would still say there are 300 pages, but you would be unable to turn to page 250.

The frame works the same way. If the frame header is incorrect (such as if a download was interrupted before it was finished), then you want to be able to tell the user what is wrong, not just crash the whole application. And crashing is what happens when you reach out of bounds into memory space that isn’t yours.

> ```swift
> let parsedSubframes = parse(
>   embeddedSubframes: frameData,
>   id3Tag: id3Tag,
>   frameSizeParser: frameSizeParser,
>   id3FrameParser: id3FrameParser)
>   // how do I get a [FrameName: ID3Frame] return from this?
> ```

I it looks like that parses `frameData` _into_ `id3Tag`, so you’re putting it into the main tag. You would need to create your own `ID3Tag` instance and pass that to `id3Tag` instead of the global one. Then you can do whatever you want with it after `parse(...)` has filled it up with the subframes.

Reply:
> The frame works the same way. If the frame header is incorrect (such as if a download was interrupted before it was finished), then you want to be able to tell the user what is wrong, not just crash the whole application. And crashing is what happens when you reach out of bounds into memory space that isn’t yours.

Okay, I understand that, and I could handle an error, probably by using the `CorruptedFile` error in `ID3TagEditorError`. Are you just referring to the need for error handling, or are you saying I need some way to validate the results of what `ID3FrameSizeParser` is returning with?

Reply:
I’m only saying you shouldn’t use an index in a collection without first knowing it is valid. For example, the `extract` methods I wrote stop either when they run out of data, or when the frame itself says to stop, whichever happens first. That way they never reach for data that doesn’t exist. Due to the lack of an error propagation mechanism in upstream’s code, they don’t bother complaining that the frame lied about its length, they just report back whatever was actually there to read. You can decide whatever you want to do if the frame lies. You just don’t want to blindly trust it, because you’ll risk crashing—or worse, [being open to attack](https://www.ssbwiki.com/Smash_Stack).